components/role-mgt/org.wso2.carbon.identity.role.v2.mgt.core/src/main/java/org/wso2/carbon/identity/role/v2/mgt/core/RoleManagementServiceImpl.java

```diff
@@ -0,0 +1,772 @@
+            throw new IdentityRoleManagementServerException(UNEXPECTED_SERVER_ERROR.getCode(), errorMessage, e);
+        }
+    }
+
+    /**
+     * Validate permissions.
+     *
+     * @param permissions Permissions.
+     * @param audience  Audience.
+     * @param audienceId  Audience ID.
+     * @param tenantDomain Tenant domain.
+     * @throws IdentityRoleManagementException Error occurred while validating permissions.
+     */
+    private void validatePermissions(List<Permission> permissions, String audience, String audienceId,
+                                     String tenantDomain)
+            throws IdentityRoleManagementException {
+
+        switch (audience) {
+            case ORGANIZATION:
+                validatePermissionsForOrganization(permissions, tenantDomain);
+                break;
+            default:
+                break;
+        }
+    }
+
+    /**
+     * Validate permissions for organization audience.
+     *
+     * @param permissions Permissions.
+     * @throws IdentityRoleManagementException Error occurred while validating permissions.
+     */
+    private void validatePermissionsForOrganization(List<Permission> permissions, String tenantDomain)
+            throws IdentityRoleManagementException {
+
+        try {
+            List<Scope> scopes = RoleManagementServiceComponentHolder.getInstance()
+                    .getApiResourceManager().getScopesByTenantDomain(tenantDomain, "");
+            List<String> scopeNameList = new ArrayList<>();
+            for (Scope scope : scopes) {
+                scopeNameList.add(scope.getName());
+            }
+            for (Permission permission : permissions) {
+
+                if (!scopeNameList.contains(permission.getName())) {
+                    throw new IdentityRoleManagementClientException(INVALID_PERMISSION.getCode(),
+                            "Permission: " + permission.getName() + " not found");
+                }
+            }
+        } catch (APIResourceMgtException e) {
+            throw new IdentityRoleManagementException("Error while retrieving scopes", "Error while retrieving scopes "
+                    + "for tenantDomain: " + tenantDomain, e);
+        }
+    }
+
+    /**
+     * Check if the role name has a domain separator character.
+     * @param roleName Role name.
+     * @return True if the role name has a domain separator character.
+     */
+    private boolean isDomainSeparatorPresent(String roleName) {
+
+        return roleName.contains(UserCoreConstants.DOMAIN_SEPARATOR);
+    }
+
+    /**
+     * Remove similar permissions.
+     *
+     * @param arr1 Array of permissions.
+     * @param arr2 Array of permissions.
+     */
+    private void removeSimilarPermissions(List<Permission> arr1, List<Permission> arr2) {
+        List<Permission> toRemove = new ArrayList<>();
+
+        for (Permission p1 : arr1) {
```

Comment:
Better override the equals and hashCode methods of the Permission class. Then comparisons will be easy.

Reply:
Same can be applicable for IdpGroup

Reply:
fixed